src/api/app/controllers/person_controller.rb

```diff
@@ -162,25 +162,14 @@ def internal_register
 
   def update_watchlist( user, xml )
     new_watchlist = []
-    old_watchlist = []
-
     xml.get('watchlist').elements("project") do |e|
       new_watchlist << e['name']
     end
 
-    user.watched_projects.each do |wp|
-      old_watchlist << wp.project.name
-    end
-    add_to_watchlist = new_watchlist.collect {|i| old_watchlist.include?(i) ? nil : i}.compact
-    remove_from_watchlist = old_watchlist.collect {|i| new_watchlist.include?(i) ? nil : i}.compact
-
-    remove_from_watchlist.each do |name|
-      user.watched_projects.where(project_id: Project.find_by_name(name).id).delete_all
-    end
-
-    add_to_watchlist.each do |name|
-      user.watched_projects.new(project_id: Project.find_by_name(name).id)
+    new_watchlist.map! do |name|
+      WatchedProject.find_or_create_by(project: Project.find_by_name!(name), user: user)
     end
+    user.watched_projects.replace(new_watchlist)
```

Comment:
I checked the docs, replace deletes projects which are not included! So the implementation looks fine!

See http://www.rubydoc.info/docs/rails/4.1.7/ActiveRecord%2FAssociations%2FCollectionProxy%3Areplace